redsky/event_streams.py

```diff
@@ -0,0 +1,711 @@
+        if input_info is None:
+            input_info = {}
+        self.outbound_descriptor_uid = None
+        self.md = md
+        self.output_info = output_info
+        self.input_info = input_info
+
+        # TODO: need multiple counters for multiple descriptors
+        # This will need to be a dict with keys of descriptor names
+        self.i = None
+        self.run_start_uid = None
+        self.provenance = {}
+        self.event_failed = False
+
+        # If the stream number is not specified its zero
+        for k, v in input_info.items():
+            if len(v) < 2 or isinstance(v, str):
+                input_info[k] = (v, 0)
+            if isinstance(v[1], Stream):
+                input_info[k] = (v[0], self.children.index(v[1]))
+
+    def emit(self, x):
+        """ Push data into the stream at this point
+
+        This is typically done only at source Streams but can theoretically be
+        done at any point
+        """
+        if x is not None:
+            result = []
+            for parent in self.parents:
+                r = parent.update(x, who=self)
+                if type(r) is list:
+                    result.extend(r)
+                else:
+                    result.append(r)
+            return [element for element in result if element is not None]
+
+    def dispatch(self, nds):
+        """Call method which correlates to the input document
+
+        Parameters
+        ----------
+        nds: tuple
+            Name document pair
+
+        Returns
+        -------
+        tuple:
+            New name document pair
+        """
+        name, docs = self.curate_streams(nds)
+        return getattr(self, name)(docs)
+
+    def update(self, x, who=None):
+        """Emit the new name document pair
+
+        Parameters
+        ----------
+        x: tuple
+            Name document pair
+        who: stream instance, optional
+            This is mostly used internally for emit
+
+        Returns
+        -------
+        list:
+            The list of issued documents, used for back pressure
+        """
+        return self.emit(self.dispatch(x))
+
+    def curate_streams(self, nds):
+        """Standardize name document pairs
+
+        Parameters
+        ----------
+        nds: tuple, or tuple of tuples
+            The name document pair(s)
+
+        Returns
+        -------
+        name: str
+            The name of the output doc(s)
+        docs: tuple
+            The document(s)
+        """
+        # If we get multiple streams make (doc, doc, doc, ...)
+        if isinstance(nds[0], tuple):
+            names, docs = list(zzip(*nds))
+            if len(set(names)) > 1:
+                raise RuntimeError('Misaligned Streams')
+            name = names[0]
+        # If only one stream then (doc, )
+        else:
+            names, docs = nds
+            name = names
+            docs = (docs,)
+        return name, docs
+
+    def generate_provenance(self, func=None):
```

Comment:
Thanks. I'm not certain if this is enough info to reproduce the pipeline from the start alone, but I can always add more stuff to it later as needed (and this PR is decidely not dealing with the database insert/retrieval/resurrect my pipeline problem for that stuff see #14).